### PR TITLE
Fix crash and record diagnostic location in preview feature analyzer

### LINKED ISSUE #5802: CSharpDetectPreviewFeatureAnalyzer hitting NullReferenceException when run on Corelib
### Analyzer

**Diagnostic ID**: [CA2252](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2252): `Opt in to preview features`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 6.0.3 (Latest)

### Describe the bug

e.g. AD0001 fired with the exception from the analyzer:
```
Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Message	AD0001	Analyzer 'Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.
Exception occurred with following context:
Compilation: System.Private.CoreLib
ISymbol: Decimal (NamedType)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer.GetPreviewInterfaceNodeForTypeImplementingPreviewInterface(ISymbol typeSymbol, ISymbol previewInterfaceSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.ProcessTypeSymbolAttributes(SymbolAnalysisContext context, ITypeSymbol symbol, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, IFieldSymbol virtualStaticsInInterfaces, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.<>c__DisplayClass33_0.<Initialize>b__2(SymbolAnalysisContext context)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c.<ExecuteSymbolActionsCore>b__50_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----

Suppress the following diagnostics to disable this analyzer: CA2252	System.Private.CoreLib		1	Active	Analyzer 'Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer' threw the following exception:
'Exception occurred with following context:
Compilation: System.Private.CoreLib
ISymbol: Decimal (NamedType)

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.NetCore.CSharp.Analyzers.Runtime.CSharpDetectPreviewFeatureAnalyzer.GetPreviewInterfaceNodeForTypeImplementingPreviewInterface(ISymbol typeSymbol, ISymbol previewInterfaceSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.ProcessTypeSymbolAttributes(SymbolAnalysisContext context, ITypeSymbol symbol, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary`2 requiresPreviewFeaturesSymbols, IFieldSymbol virtualStaticsInInterfaces, INamedTypeSymbol previewFeatureAttributeSymbol)
   at Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer.<>c__DisplayClass33_0.<Initialize>b__2(SymbolAnalysisContext context)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c.<ExecuteSymbolActionsCore>b__50_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----
'.
```
cc: @jeffhandley

### PR DESCRIPTION
**Pull request by Youssef1313**
@jeffhandley Can you take a look?

Fixes #5802

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs [hunk 1]
```diff
@@ -942,7 +942,7 @@ private static ISymbol SymbolFromArrayCreationOperation(IArrayCreationOperation
         {
             if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out (bool isPreview, string? message, string? url) existing))
             {
-                Debug.Assert(true, $"Should never reach this line. This means the symbol {symbol.Name} was not processed in this analyzer");
+                Debug.Assert(false, $"Should never reach this line. This means the symbol {symbol.Name} was not processed in this analyzer");
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:945]: This looks like a typo. Previously, the assert will always pass and we wouldn't know there is an issue.

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:945]: Yep looks weird. I would replace `Debug.Assert(false` by `Debug.Fail` which is clearer.

**Review by buyaa-n — CHANGES_REQUESTED** [review]
Thanks for the fix @Youssef1313 overall LGTM , could you add similar logic into VB code? Looks at least this method could benefit fom using `TypeStatementSyntax` and `TypeBlockSyntax` instead
https://github.com/dotnet/roslyn-analyzers/blob/f471d3381584f10f9908432e0b2b2b8ef07a0aa6/src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Runtime/BasicDetectPreviewFeatureAnalyzer.vb#L124-L176

**Review by buyaa-n — APPROVED** [review]
Thanks @Youssef1313 looks great. The perf failures unrelated and reported in https://github.com/dotnet/roslyn-analyzers/issues/5790, merging

**Pull request merged**